### PR TITLE
Extended JSON support

### PR DESCRIPTION
**Pull request by Tekki**
As discussed in https://github.com/Tekki/mojo-mysql/issues/1

### CODE REVIEW

**Comment by jhthorsen** [comment]
Hey! I think this is a great extension, but I do have some questions: Is any JSON document supported? Like what about strings (not sure if that makes sense), but at least ARRAY on the top level? I wonder if it would be safer to do: `{column => {json => $whatever}}`. I know it's a bit more verbose, but at least it will always work.

The whole thing about inserting documents is a bit confusing to me, since I thought that was already handled by Mojo::mysql::Database.

I also see some commas that is not required, like `,}`, but I can clean that up later on...

I'll have a closer look tomorrow, but this will get merged on way or another 👍

**Comment by Tekki** [comment]
We look at the things from a different point of view.
- `,` at the end of lists is a question of style and the style in this repo is defined by you
- think of a document as a complete row and not just as values of one column; first dimension of keys are the columns, deeper dimensions are documents inside columns
- forget about 'JSON'; how documents are implemented in a database is a technical question that a user of an abstract module shouldn't be confronted with
- the abstract module should work standalone and not just in conjunction with Database
- there are no tests for inserting an array at the first level, so there's a chance it doesn't work yet

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm]: Is there a downside to supporting the same syntax as in https://github.com/jhthorsen/mojo-mysql/blob/38c16bdc6822e2039a258b46170e3874541fa921/lib/Mojo/mysql/Database.pm#L311

Like this:

```
if (ref $v eq 'HASH' and exists $param->{json}) { ... }
```

It's a bit extra to type, but it would allow it to be extended further in the future.

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm]: If you want to be "git blamed" for these lines, then you have to remove the extra comma before `}`.

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm]: Could you compress this to:

```
sub _order_by_chunks {
  my ($self, $arg) = @_;
  return $self->_json_extract($arg) if $arg =~ /->/;
  return $self->SUPER::_order_by_chunks($arg);
}
```

Same with `_select_fields()` below.

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm]: I'm not a fan of the extra `->`. Please write it as `$data->{$label}[0]`. (Seen other places in this PR as well)

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm:158]: Is there a test for when this regex fails to match `$k`? I'm unsure from reading the code about what happens if `$k` is simply "foo", meaning without `->`.

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm:153]: I think this whole method is a bit hard to read, because of all the nested if/else. Can you move some of the code out in methods with descriptive names? I think that could limit the nesting by using `return` instead.

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm:174]: Is this symmetric with what you have at https://github.com/jhthorsen/mojo-mysql/pull/51/files#diff-45bdaf6ecc6b539751f37122c23c1f17R53 ?

`if (ref $v eq 'HASH') { ...`

File: lib/SQL/Abstract/mysql.pm
Comment (jhthorsen) [inline lib/SQL/Abstract/mysql.pm:214]: This method could be compressed to something like:

```
$k = \$self->_json_extract($k) if $k =~ /->/;
return $self->SUPER::_where_hashpair_SCALAR($k, $v);
```

**Review by jhthorsen — CHANGES_REQUESTED** [review]
Sorry for all the nitpicking, but there are some parts of this code that I have a hard time wrapping my head around. However, the documentation seems good and there's a bunch of new tests, so I'm still very much positive.

Another thing: What about `$obj->TO_JSON`? I think that should be supported as well.

Maybe @shadowcat-mst has some input on the syntax? Especially regarding `{column => \%json}` v.s `{column => {json => \%json}}`.

**Comment by Tekki** [comment]
I've no problems with your nitpicking and there is no hurry for this new feature.
In general I'm not too happy about the way I've implemented the changes. SQL::Abstract gives us a lot of possibilities to override internal methods and I tried not to disable too much of the original code. But unfortunately most of the time this possibility is missing, so I had to use maybe 3 different approaches.

If you still ask about `column => {json => $whatever}` it is because you didn't forget about JSON. I give you an example: Let's say we have a table with columns `firstname`, `lastname` and `address`. From the API you get a POST request with
```
$request = {
  firstname => 'Haymitch',
  lastname => 'Abernathy',
  address => {place => 'Victors Village', district => 12}
}
```
Now what to expect from an abstract module is that it knows how to handle `->insert($table, $request)` so you don't have to split it into
```
{
  firstname => $request->{firstname},
  lastname => $request->{lastname},
  address => {json => $request->{address}}
}
```
As I said earlier, the user should't have to worry about the technical way how the database stores a hash reference.

File: lib/SQL/Abstract/mysql.pm
Comment (Tekki) [reply lib/SQL/Abstract/mysql.pm]: A shorter answer than the one below: We don't write
`$db->query('insert into foo values (?)', {integer => 23});` either.

File: lib/SQL/Abstract/mysql.pm
Comment (Tekki) [reply lib/SQL/Abstract/mysql.pm:158]: If the key is "foo" it will not appear in `$k` because of the grep. This block only processes keys with arrows.

**Comment by Tekki** [comment]
Tried to discuss this topic too at the Zurich Perl Meetup. But the discussion immediately went to the question why we should use JSON at all. We were looking for real use cases and until we found one (postal addresses) the time was over.

File: lib/SQL/Abstract/mysql.pm
Comment (Tekki) [reply lib/SQL/Abstract/mysql.pm:174]: It's not symmetric and `if (ref $v eq 'HASH') { ...` is wrong because it doesn't accept array references (the top level of a document can be an array, as you have mentioned in your first comment).
At this place we process the value of `field->some.key` and we can expect that everything that is a reference has to be JSON encoded.

File: lib/SQL/Abstract/mysql.pm
Comment (Tekki) [reply lib/SQL/Abstract/mysql.pm:153]: It's hard to read, but there are too many variables for a method and the block is only 40 lines long. Maybe some comments will help?

**Comment by jhthorsen** [comment]
@Tekki: I think this is a great enhancement and it also resolves #57. Any way you can resolve the conflict and (force) push other cleanups?

If you want, we can mark this feature as EXPERIMENTAL, even though I don't think we'll get rid of it once it's merged. I do like "EXPERIMENTAL" though, in case there's parts that need tweaking or if there is a big chance for bugs.

**Comment by Tekki** [comment]
I have to take a look at it again, since we made several changes in the meantime.

**Comment by jhthorsen** [comment]
I'm closing this for now, since it seems to have gone stale. Please do create a new branch in this repo if you want to complete it, now that you have commit bit.

**Comment by Tekki** [comment]
It has gone stale because we had other things to do. But the code has changed a lot in the meantime and this needs a new approach, so it's okay to close it.

**Pull request closed without merging**
